src/gcp-opt.jl

```diff
@@ -174,21 +176,33 @@ function _gcp(
     return CPD(λ, Tuple(U))
 end
 
-# inefficient but simple
+# Faster MTTKRP
 function mttkrp(X, U, n)
+
     # Dimensions
     N, I, r = length(U), Tuple(size.(U, 1)), (only∘unique)(size.(U, 2))
     (N == ndims(X) && I == size(X)) || throw(DimensionMismatch("`X` and `U` do not have matching dimensions"))
 
-    # Matricized tensor (in mode n)
-    Xn = reshape(permutedims(X, [n; setdiff(1:N, n)]), size(X, n), :)
-
-    # Khatri-Rao product (in mode n)
-    Zn = similar(U[1], prod(I[setdiff(1:N, n)]), r)
+    # See section III-B from "Fast Alternating LS Algorithms for High Order CANDECOMP/PARAFAC Tensor Factorizations" by Phan et al.
+    Rn = similar(U[n])
+    Jn = prod(size(X)[1:n])
+    Kn = prod(size(X)[n+1:end])
+    # Compute tensor-vector products right to left (equations 15, 17) for each rank
     for j in 1:r
-        Zn[:, j] = reduce(kron, [view(U[i], :, j) for i in reverse(setdiff(1:N, n))])
+        # Inner tensor-vector products
+        Rn_j = reshape(reshape(X, Jn, Kn) * reduce(kron, [view(U[i], :, j) for i in reverse(n+1:N)], init=1), size(X)[1:n]) 
+        
+        # Outer tensor-vector products
+        # Permute so dims to be multiplied are last
+        Rn_j = permutedims(Rn_j, [n; 1:n-1])
+        # Multiply from right to left
+        sz = size(Rn_j)
+        m = length(sz)
+        for k in n-1:-1:1
+            Rn_j = reshape(Rn_j, prod(sz[1:m-1]), sz[m]) * U[k][:, j]
+            m -= 1
+        end
+        Rn[:, j] = Rn_j  
```

Comment:
You should profile to double check, but my guess is that the slow-down is coming from:
+ the `permutedims` on Line 197
+ the matrix multiplies on Line 202
Both involve significant allocation. BTW note that this mttkrp algorithms is largely about trying to avoid `permutedims`!

There's actually a way to write this as a single matrix-vector product - see if you can figure it out! Hint: take a look at the inner tensor-vector product from (14) and note the similarity to the outer tensor-vector product you want.